# Problem Statement
## Prepare rules for the all the data sets 
### Try different values of support and confidence. Observe the change in number of rules for different support,confidence values
### Change the minimum length in apriori algorithm
### Visulize the obtained rules using different plots 

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from mlxtend.frequent_patterns import apriori,association_rules
from mlxtend.preprocessing import TransactionEncoder

In [2]:
df = pd.read_csv(r'/Users/ag/Downloads/my_movies.csv')

In [3]:
df.head()

V1       V2             V3          V4     V5  Sixth Sense  \
0  Sixth Sense    LOTR1  Harry Potter1  Green Mile  LOTR2            1   
1    Gladiator  Patriot     Braveheart         NaN    NaN            0   
2        LOTR1    LOTR2            NaN         NaN    NaN            0   
3    Gladiator  Patriot    Sixth Sense         NaN    NaN            1   
4    Gladiator  Patriot    Sixth Sense         NaN    NaN            1   

   Gladiator  LOTR1  Harry Potter1  Patriot  LOTR2  Harry Potter2  LOTR  \
0          0      1              1        0      1              0     0   
1          1      0              0        1      0              0     0   
2          0      1              0        0      1              0     0   
3          1      0              0        1      0              0     0   
4          1      0              0        1      0              0     0   

   Braveheart  Green Mile  
0           0           1  
1           1           0  
2           0           0  
3           0           0  
4           0           0

In [4]:
df.shape

(10, 15)

In [5]:
df.isnull().sum()

V1               0
V2               0
V3               3
V4               8
V5               9
Sixth Sense      0
Gladiator        0
LOTR1            0
Harry Potter1    0
Patriot          0
LOTR2            0
Harry Potter2    0
LOTR             0
Braveheart       0
Green Mile       0
dtype: int64

In [6]:
# removing columns v1,v2,v3,v4,v5 as pd.dummies is already applied on the dataset
df.drop(columns=['V1','V2','V3','V4','V5'], inplace=True)

In [7]:
df

Sixth Sense  Gladiator  LOTR1  Harry Potter1  Patriot  LOTR2  \
0            1          0      1              1        0      1   
1            0          1      0              0        1      0   
2            0          0      1              0        0      1   
3            1          1      0              0        1      0   
4            1          1      0              0        1      0   
5            1          1      0              0        1      0   
6            0          0      0              1        0      0   
7            0          1      0              0        1      0   
8            1          1      0              0        1      0   
9            1          1      0              0        0      0   

   Harry Potter2  LOTR  Braveheart  Green Mile  
0              0     0           0           1  
1              0     0           1           0  
2              0     0           0           0  
3              0     0           0           0  
4              0     0           0           0  
5              0     0           0           0  
6              1     0           0           0  
7              0     0           0           0  
8              0     0           0           0  
9              0     1           0           1

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10 entries, 0 to 9
Data columns (total 10 columns):
 #   Column         Non-Null Count  Dtype
---  ------         --------------  -----
 0   Sixth Sense    10 non-null     int64
 1   Gladiator      10 non-null     int64
 2   LOTR1          10 non-null     int64
 3   Harry Potter1  10 non-null     int64
 4   Patriot        10 non-null     int64
 5   LOTR2          10 non-null     int64
 6   Harry Potter2  10 non-null     int64
 7   LOTR           10 non-null     int64
 8   Braveheart     10 non-null     int64
 9   Green Mile     10 non-null     int64
dtypes: int64(10)
memory usage: 928.0 bytes


In [9]:
df.dtypes

Sixth Sense      int64
Gladiator        int64
LOTR1            int64
Harry Potter1    int64
Patriot          int64
LOTR2            int64
Harry Potter2    int64
LOTR             int64
Braveheart       int64
Green Mile       int64
dtype: object

In [10]:
frequent_itemsets = apriori(df, min_support=0.1, use_colnames=True)
frequent_itemsets

/opt/anaconda3/lib/python3.9/site-packages/mlxtend/frequent_patterns/fpcommon.py:111: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


support                                           itemsets
0       0.6                                      (Sixth Sense)
1       0.7                                        (Gladiator)
2       0.2                                            (LOTR1)
3       0.2                                    (Harry Potter1)
4       0.6                                          (Patriot)
5       0.2                                            (LOTR2)
6       0.1                                    (Harry Potter2)
7       0.1                                             (LOTR)
8       0.1                                       (Braveheart)
9       0.2                                       (Green Mile)
10      0.5                           (Gladiator, Sixth Sense)
11      0.1                               (LOTR1, Sixth Sense)
12      0.1                       (Harry Potter1, Sixth Sense)
13      0.4                             (Patriot, Sixth Sense)
14      0.1                               (Sixth Sense, LOTR2)
15      0.1                                (LOTR, Sixth Sense)
16      0.2                          (Sixth Sense, Green Mile)
17      0.6                               (Patriot, Gladiator)
18      0.1                                  (LOTR, Gladiator)
19      0.1                            (Braveheart, Gladiator)
20      0.1                            (Gladiator, Green Mile)
21      0.1                             (Harry Potter1, LOTR1)
22      0.2                                     (LOTR1, LOTR2)
23      0.1                                (LOTR1, Green Mile)
24      0.1                             (Harry Potter1, LOTR2)
25      0.1                     (Harry Potter1, Harry Potter2)
26      0.1                        (Harry Potter1, Green Mile)
27      0.1                              (Braveheart, Patriot)
28      0.1                                (LOTR2, Green Mile)
29      0.1                                 (LOTR, Green Mile)
30      0.4                  (Gladiator, Patriot, Sixth Sense)
31      0.1                     (LOTR, Gladiator, Sixth Sense)
32      0.1               (Gladiator, Sixth Sense, Green Mile)
33      0.1                (LOTR1, Harry Potter1, Sixth Sense)
34      0.1                        (LOTR1, Sixth Sense, LOTR2)
35      0.1                   (LOTR1, Sixth Sense, Green Mile)
36      0.1                (Harry Potter1, Sixth Sense, LOTR2)
37      0.1           (Harry Potter1, Sixth Sense, Green Mile)
38      0.1                   (Sixth Sense, LOTR2, Green Mile)
39      0.1                    (LOTR, Sixth Sense, Green Mile)
40      0.1                   (Braveheart, Patriot, Gladiator)
41      0.1                      (LOTR, Gladiator, Green Mile)
42      0.1                      (Harry Potter1, LOTR1, LOTR2)
43      0.1                 (Harry Potter1, LOTR1, Green Mile)
44      0.1                         (LOTR1, LOTR2, Green Mile)
45      0.1                 (Harry Potter1, LOTR2, Green Mile)
46      0.1         (LOTR, Gladiator, Sixth Sense, Green Mile)
47      0.1         (LOTR1, Harry Potter1, Sixth Sense, LOTR2)
48      0.1    (LOTR1, Harry Potter1, Sixth Sense, Green Mile)
49      0.1            (LOTR1, Sixth Sense, LOTR2, Green Mile)
50      0.1    (Harry Potter1, Sixth Sense, LOTR2, Green Mile)
51      0.1          (Harry Potter1, LOTR1, LOTR2, Green Mile)
52      0.1  (Harry Potter1, LOTR1, Sixth Sense, LOTR2, Gre...

In [11]:
rules = association_rules(frequent_itemsets, metric="lift", min_threshold=0.7)
rules

antecedents                                      consequents  \
0        (Gladiator)                                    (Sixth Sense)   
1      (Sixth Sense)                                      (Gladiator)   
2            (LOTR1)                                    (Sixth Sense)   
3      (Sixth Sense)                                          (LOTR1)   
4    (Harry Potter1)                                    (Sixth Sense)   
..               ...                                              ...   
245  (Harry Potter1)          (Sixth Sense, LOTR1, LOTR2, Green Mile)   
246          (LOTR1)  (Harry Potter1, Sixth Sense, LOTR2, Green Mile)   
247    (Sixth Sense)        (Harry Potter1, LOTR1, LOTR2, Green Mile)   
248          (LOTR2)  (Sixth Sense, Harry Potter1, LOTR1, Green Mile)   
249     (Green Mile)       (Sixth Sense, Harry Potter1, LOTR1, LOTR2)   

     antecedent support  consequent support  support  confidence      lift  \
0                   0.7                 0.6      0.5    0.714286  1.190476   
1                   0.6                 0.7      0.5    0.833333  1.190476   
2                   0.2                 0.6      0.1    0.500000  0.833333   
3                   0.6                 0.2      0.1    0.166667  0.833333   
4                   0.2                 0.6      0.1    0.500000  0.833333   
..                  ...                 ...      ...         ...       ...   
245                 0.2                 0.1      0.1    0.500000  5.000000   
246                 0.2                 0.1      0.1    0.500000  5.000000   
247                 0.6                 0.1      0.1    0.166667  1.666667   
248                 0.2                 0.1      0.1    0.500000  5.000000   
249                 0.2                 0.1      0.1    0.500000  5.000000   

     leverage  conviction  
0        0.08        1.40  
1        0.08        1.80  
2       -0.02        0.80  
3       -0.02        0.96  
4       -0.02        0.80  
..        ...         ...  
245      0.08        1.80  
246      0.08        1.80  
247      0.04        1.08  
248      0.08        1.80  
249      0.08        1.80  

[250 rows x 9 columns]

In [12]:
rules0_1 = association_rules(frequent_itemsets, metric="lift", min_threshold=0.1)
rules0_1

antecedents                                      consequents  \
0        (Gladiator)                                    (Sixth Sense)   
1      (Sixth Sense)                                      (Gladiator)   
2            (LOTR1)                                    (Sixth Sense)   
3      (Sixth Sense)                                          (LOTR1)   
4    (Harry Potter1)                                    (Sixth Sense)   
..               ...                                              ...   
245  (Harry Potter1)          (Sixth Sense, LOTR1, LOTR2, Green Mile)   
246          (LOTR1)  (Harry Potter1, Sixth Sense, LOTR2, Green Mile)   
247    (Sixth Sense)        (Harry Potter1, LOTR1, LOTR2, Green Mile)   
248          (LOTR2)  (Sixth Sense, Harry Potter1, LOTR1, Green Mile)   
249     (Green Mile)       (Sixth Sense, Harry Potter1, LOTR1, LOTR2)   

     antecedent support  consequent support  support  confidence      lift  \
0                   0.7                 0.6      0.5    0.714286  1.190476   
1                   0.6                 0.7      0.5    0.833333  1.190476   
2                   0.2                 0.6      0.1    0.500000  0.833333   
3                   0.6                 0.2      0.1    0.166667  0.833333   
4                   0.2                 0.6      0.1    0.500000  0.833333   
..                  ...                 ...      ...         ...       ...   
245                 0.2                 0.1      0.1    0.500000  5.000000   
246                 0.2                 0.1      0.1    0.500000  5.000000   
247                 0.6                 0.1      0.1    0.166667  1.666667   
248                 0.2                 0.1      0.1    0.500000  5.000000   
249                 0.2                 0.1      0.1    0.500000  5.000000   

     leverage  conviction  
0        0.08        1.40  
1        0.08        1.80  
2       -0.02        0.80  
3       -0.02        0.96  
4       -0.02        0.80  
..        ...         ...  
245      0.08        1.80  
246      0.08        1.80  
247      0.04        1.08  
248      0.08        1.80  
249      0.08        1.80  

[250 rows x 9 columns]

In [13]:
rules0_2 = association_rules(frequent_itemsets, metric="lift", min_threshold=0.2)
rules0_2

antecedents                                      consequents  \
0        (Gladiator)                                    (Sixth Sense)   
1      (Sixth Sense)                                      (Gladiator)   
2            (LOTR1)                                    (Sixth Sense)   
3      (Sixth Sense)                                          (LOTR1)   
4    (Harry Potter1)                                    (Sixth Sense)   
..               ...                                              ...   
245  (Harry Potter1)          (Sixth Sense, LOTR1, LOTR2, Green Mile)   
246          (LOTR1)  (Harry Potter1, Sixth Sense, LOTR2, Green Mile)   
247    (Sixth Sense)        (Harry Potter1, LOTR1, LOTR2, Green Mile)   
248          (LOTR2)  (Sixth Sense, Harry Potter1, LOTR1, Green Mile)   
249     (Green Mile)       (Sixth Sense, Harry Potter1, LOTR1, LOTR2)   

     antecedent support  consequent support  support  confidence      lift  \
0                   0.7                 0.6      0.5    0.714286  1.190476   
1                   0.6                 0.7      0.5    0.833333  1.190476   
2                   0.2                 0.6      0.1    0.500000  0.833333   
3                   0.6                 0.2      0.1    0.166667  0.833333   
4                   0.2                 0.6      0.1    0.500000  0.833333   
..                  ...                 ...      ...         ...       ...   
245                 0.2                 0.1      0.1    0.500000  5.000000   
246                 0.2                 0.1      0.1    0.500000  5.000000   
247                 0.6                 0.1      0.1    0.166667  1.666667   
248                 0.2                 0.1      0.1    0.500000  5.000000   
249                 0.2                 0.1      0.1    0.500000  5.000000   

     leverage  conviction  
0        0.08        1.40  
1        0.08        1.80  
2       -0.02        0.80  
3       -0.02        0.96  
4       -0.02        0.80  
..        ...         ...  
245      0.08        1.80  
246      0.08        1.80  
247      0.04        1.08  
248      0.08        1.80  
249      0.08        1.80  

[250 rows x 9 columns]

In [14]:
# Creating rules with Min_support = 0.2
frequent_itemsets1 = apriori(df, min_support=0.2, use_colnames=True)
frequent_itemsets1

/opt/anaconda3/lib/python3.9/site-packages/mlxtend/frequent_patterns/fpcommon.py:111: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


support                           itemsets
0       0.6                      (Sixth Sense)
1       0.7                        (Gladiator)
2       0.2                            (LOTR1)
3       0.2                    (Harry Potter1)
4       0.6                          (Patriot)
5       0.2                            (LOTR2)
6       0.2                       (Green Mile)
7       0.5           (Gladiator, Sixth Sense)
8       0.4             (Patriot, Sixth Sense)
9       0.2          (Sixth Sense, Green Mile)
10      0.6               (Patriot, Gladiator)
11      0.2                     (LOTR1, LOTR2)
12      0.4  (Gladiator, Patriot, Sixth Sense)

In [15]:
rules1 = association_rules(frequent_itemsets1, metric="lift", min_threshold=0.7)
rules1

antecedents               consequents  antecedent support  \
0                (Gladiator)             (Sixth Sense)                 0.7   
1              (Sixth Sense)               (Gladiator)                 0.6   
2                  (Patriot)             (Sixth Sense)                 0.6   
3              (Sixth Sense)                 (Patriot)                 0.6   
4              (Sixth Sense)              (Green Mile)                 0.6   
5               (Green Mile)             (Sixth Sense)                 0.2   
6                  (Patriot)               (Gladiator)                 0.6   
7                (Gladiator)                 (Patriot)                 0.7   
8                    (LOTR1)                   (LOTR2)                 0.2   
9                    (LOTR2)                   (LOTR1)                 0.2   
10      (Patriot, Gladiator)             (Sixth Sense)                 0.6   
11  (Sixth Sense, Gladiator)                 (Patriot)                 0.5   
12    (Patriot, Sixth Sense)               (Gladiator)                 0.4   
13               (Gladiator)    (Patriot, Sixth Sense)                 0.7   
14                 (Patriot)  (Sixth Sense, Gladiator)                 0.6   
15             (Sixth Sense)      (Patriot, Gladiator)                 0.6   

    consequent support  support  confidence      lift  leverage  conviction  
0                  0.6      0.5    0.714286  1.190476      0.08         1.4  
1                  0.7      0.5    0.833333  1.190476      0.08         1.8  
2                  0.6      0.4    0.666667  1.111111      0.04         1.2  
3                  0.6      0.4    0.666667  1.111111      0.04         1.2  
4                  0.2      0.2    0.333333  1.666667      0.08         1.2  
5                  0.6      0.2    1.000000  1.666667      0.08         inf  
6                  0.7      0.6    1.000000  1.428571      0.18         inf  
7                  0.6      0.6    0.857143  1.428571      0.18         2.8  
8                  0.2      0.2    1.000000  5.000000      0.16         inf  
9                  0.2      0.2    1.000000  5.000000      0.16         inf  
10                 0.6      0.4    0.666667  1.111111      0.04         1.2  
11                 0.6      0.4    0.800000  1.333333      0.10         2.0  
12                 0.7      0.4    1.000000  1.428571      0.12         inf  
13                 0.4      0.4    0.571429  1.428571      0.12         1.4  
14                 0.5      0.4    0.666667  1.333333      0.10         1.5  
15                 0.6      0.4    0.666667  1.111111      0.04         1.2

In [16]:
rules1_1 = association_rules(frequent_itemsets1, metric="lift", min_threshold=0.1)
rules1_1

antecedents               consequents  antecedent support  \
0                (Gladiator)             (Sixth Sense)                 0.7   
1              (Sixth Sense)               (Gladiator)                 0.6   
2                  (Patriot)             (Sixth Sense)                 0.6   
3              (Sixth Sense)                 (Patriot)                 0.6   
4              (Sixth Sense)              (Green Mile)                 0.6   
5               (Green Mile)             (Sixth Sense)                 0.2   
6                  (Patriot)               (Gladiator)                 0.6   
7                (Gladiator)                 (Patriot)                 0.7   
8                    (LOTR1)                   (LOTR2)                 0.2   
9                    (LOTR2)                   (LOTR1)                 0.2   
10      (Patriot, Gladiator)             (Sixth Sense)                 0.6   
11  (Sixth Sense, Gladiator)                 (Patriot)                 0.5   
12    (Patriot, Sixth Sense)               (Gladiator)                 0.4   
13               (Gladiator)    (Patriot, Sixth Sense)                 0.7   
14                 (Patriot)  (Sixth Sense, Gladiator)                 0.6   
15             (Sixth Sense)      (Patriot, Gladiator)                 0.6   

    consequent support  support  confidence      lift  leverage  conviction  
0                  0.6      0.5    0.714286  1.190476      0.08         1.4  
1                  0.7      0.5    0.833333  1.190476      0.08         1.8  
2                  0.6      0.4    0.666667  1.111111      0.04         1.2  
3                  0.6      0.4    0.666667  1.111111      0.04         1.2  
4                  0.2      0.2    0.333333  1.666667      0.08         1.2  
5                  0.6      0.2    1.000000  1.666667      0.08         inf  
6                  0.7      0.6    1.000000  1.428571      0.18         inf  
7                  0.6      0.6    0.857143  1.428571      0.18         2.8  
8                  0.2      0.2    1.000000  5.000000      0.16         inf  
9                  0.2      0.2    1.000000  5.000000      0.16         inf  
10                 0.6      0.4    0.666667  1.111111      0.04         1.2  
11                 0.6      0.4    0.800000  1.333333      0.10         2.0  
12                 0.7      0.4    1.000000  1.428571      0.12         inf  
13                 0.4      0.4    0.571429  1.428571      0.12         1.4  
14                 0.5      0.4    0.666667  1.333333      0.10         1.5  
15                 0.6      0.4    0.666667  1.111111      0.04         1.2

In [17]:
rules1_2 = association_rules(frequent_itemsets1, metric="lift", min_threshold=0.2)
rules1_2

antecedents               consequents  antecedent support  \
0                (Gladiator)             (Sixth Sense)                 0.7   
1              (Sixth Sense)               (Gladiator)                 0.6   
2                  (Patriot)             (Sixth Sense)                 0.6   
3              (Sixth Sense)                 (Patriot)                 0.6   
4              (Sixth Sense)              (Green Mile)                 0.6   
5               (Green Mile)             (Sixth Sense)                 0.2   
6                  (Patriot)               (Gladiator)                 0.6   
7                (Gladiator)                 (Patriot)                 0.7   
8                    (LOTR1)                   (LOTR2)                 0.2   
9                    (LOTR2)                   (LOTR1)                 0.2   
10      (Patriot, Gladiator)             (Sixth Sense)                 0.6   
11  (Sixth Sense, Gladiator)                 (Patriot)                 0.5   
12    (Patriot, Sixth Sense)               (Gladiator)                 0.4   
13               (Gladiator)    (Patriot, Sixth Sense)                 0.7   
14                 (Patriot)  (Sixth Sense, Gladiator)                 0.6   
15             (Sixth Sense)      (Patriot, Gladiator)                 0.6   

    consequent support  support  confidence      lift  leverage  conviction  
0                  0.6      0.5    0.714286  1.190476      0.08         1.4  
1                  0.7      0.5    0.833333  1.190476      0.08         1.8  
2                  0.6      0.4    0.666667  1.111111      0.04         1.2  
3                  0.6      0.4    0.666667  1.111111      0.04         1.2  
4                  0.2      0.2    0.333333  1.666667      0.08         1.2  
5                  0.6      0.2    1.000000  1.666667      0.08         inf  
6                  0.7      0.6    1.000000  1.428571      0.18         inf  
7                  0.6      0.6    0.857143  1.428571      0.18         2.8  
8                  0.2      0.2    1.000000  5.000000      0.16         inf  
9                  0.2      0.2    1.000000  5.000000      0.16         inf  
10                 0.6      0.4    0.666667  1.111111      0.04         1.2  
11                 0.6      0.4    0.800000  1.333333      0.10         2.0  
12                 0.7      0.4    1.000000  1.428571      0.12         inf  
13                 0.4      0.4    0.571429  1.428571      0.12         1.4  
14                 0.5      0.4    0.666667  1.333333      0.10         1.5  
15                 0.6      0.4    0.666667  1.111111      0.04         1.2

In [18]:
# Creating rules with min_support = 0.3
frequent_itemsets2 = apriori(df, min_support=0.3, use_colnames=True)
frequent_itemsets2

/opt/anaconda3/lib/python3.9/site-packages/mlxtend/frequent_patterns/fpcommon.py:111: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


support                           itemsets
0      0.6                      (Sixth Sense)
1      0.7                        (Gladiator)
2      0.6                          (Patriot)
3      0.5           (Gladiator, Sixth Sense)
4      0.4             (Patriot, Sixth Sense)
5      0.6               (Patriot, Gladiator)
6      0.4  (Gladiator, Patriot, Sixth Sense)

In [19]:
rules2_1 = association_rules(frequent_itemsets2, metric="lift", min_threshold=0.1)
rules2_1

antecedents               consequents  antecedent support  \
0                (Gladiator)             (Sixth Sense)                 0.7   
1              (Sixth Sense)               (Gladiator)                 0.6   
2                  (Patriot)             (Sixth Sense)                 0.6   
3              (Sixth Sense)                 (Patriot)                 0.6   
4                  (Patriot)               (Gladiator)                 0.6   
5                (Gladiator)                 (Patriot)                 0.7   
6       (Patriot, Gladiator)             (Sixth Sense)                 0.6   
7   (Sixth Sense, Gladiator)                 (Patriot)                 0.5   
8     (Patriot, Sixth Sense)               (Gladiator)                 0.4   
9                (Gladiator)    (Patriot, Sixth Sense)                 0.7   
10                 (Patriot)  (Sixth Sense, Gladiator)                 0.6   
11             (Sixth Sense)      (Patriot, Gladiator)                 0.6   

    consequent support  support  confidence      lift  leverage  conviction  
0                  0.6      0.5    0.714286  1.190476      0.08         1.4  
1                  0.7      0.5    0.833333  1.190476      0.08         1.8  
2                  0.6      0.4    0.666667  1.111111      0.04         1.2  
3                  0.6      0.4    0.666667  1.111111      0.04         1.2  
4                  0.7      0.6    1.000000  1.428571      0.18         inf  
5                  0.6      0.6    0.857143  1.428571      0.18         2.8  
6                  0.6      0.4    0.666667  1.111111      0.04         1.2  
7                  0.6      0.4    0.800000  1.333333      0.10         2.0  
8                  0.7      0.4    1.000000  1.428571      0.12         inf  
9                  0.4      0.4    0.571429  1.428571      0.12         1.4  
10                 0.5      0.4    0.666667  1.333333      0.10         1.5  
11                 0.6      0.4    0.666667  1.111111      0.04         1.2

In [20]:
rules2_2 = association_rules(frequent_itemsets2, metric="lift", min_threshold=0.2)
rules2_2

antecedents               consequents  antecedent support  \
0                (Gladiator)             (Sixth Sense)                 0.7   
1              (Sixth Sense)               (Gladiator)                 0.6   
2                  (Patriot)             (Sixth Sense)                 0.6   
3              (Sixth Sense)                 (Patriot)                 0.6   
4                  (Patriot)               (Gladiator)                 0.6   
5                (Gladiator)                 (Patriot)                 0.7   
6       (Patriot, Gladiator)             (Sixth Sense)                 0.6   
7   (Sixth Sense, Gladiator)                 (Patriot)                 0.5   
8     (Patriot, Sixth Sense)               (Gladiator)                 0.4   
9                (Gladiator)    (Patriot, Sixth Sense)                 0.7   
10                 (Patriot)  (Sixth Sense, Gladiator)                 0.6   
11             (Sixth Sense)      (Patriot, Gladiator)                 0.6   

    consequent support  support  confidence      lift  leverage  conviction  
0                  0.6      0.5    0.714286  1.190476      0.08         1.4  
1                  0.7      0.5    0.833333  1.190476      0.08         1.8  
2                  0.6      0.4    0.666667  1.111111      0.04         1.2  
3                  0.6      0.4    0.666667  1.111111      0.04         1.2  
4                  0.7      0.6    1.000000  1.428571      0.18         inf  
5                  0.6      0.6    0.857143  1.428571      0.18         2.8  
6                  0.6      0.4    0.666667  1.111111      0.04         1.2  
7                  0.6      0.4    0.800000  1.333333      0.10         2.0  
8                  0.7      0.4    1.000000  1.428571      0.12         inf  
9                  0.4      0.4    0.571429  1.428571      0.12         1.4  
10                 0.5      0.4    0.666667  1.333333      0.10         1.5  
11                 0.6      0.4    0.666667  1.111111      0.04         1.2

In [21]:
# Creating rules with min_support = 0.4
frequent_itemsets3 = apriori(df, min_support=0.4, use_colnames=True)
frequent_itemsets3

/opt/anaconda3/lib/python3.9/site-packages/mlxtend/frequent_patterns/fpcommon.py:111: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


support                           itemsets
0      0.6                      (Sixth Sense)
1      0.7                        (Gladiator)
2      0.6                          (Patriot)
3      0.5           (Gladiator, Sixth Sense)
4      0.4             (Patriot, Sixth Sense)
5      0.6               (Patriot, Gladiator)
6      0.4  (Gladiator, Patriot, Sixth Sense)

In [22]:
rules3 = association_rules(frequent_itemsets3, metric="lift", min_threshold=0.1)
rules3

antecedents               consequents  antecedent support  \
0                (Gladiator)             (Sixth Sense)                 0.7   
1              (Sixth Sense)               (Gladiator)                 0.6   
2                  (Patriot)             (Sixth Sense)                 0.6   
3              (Sixth Sense)                 (Patriot)                 0.6   
4                  (Patriot)               (Gladiator)                 0.6   
5                (Gladiator)                 (Patriot)                 0.7   
6       (Patriot, Gladiator)             (Sixth Sense)                 0.6   
7   (Sixth Sense, Gladiator)                 (Patriot)                 0.5   
8     (Patriot, Sixth Sense)               (Gladiator)                 0.4   
9                (Gladiator)    (Patriot, Sixth Sense)                 0.7   
10                 (Patriot)  (Sixth Sense, Gladiator)                 0.6   
11             (Sixth Sense)      (Patriot, Gladiator)                 0.6   

    consequent support  support  confidence      lift  leverage  conviction  
0                  0.6      0.5    0.714286  1.190476      0.08         1.4  
1                  0.7      0.5    0.833333  1.190476      0.08         1.8  
2                  0.6      0.4    0.666667  1.111111      0.04         1.2  
3                  0.6      0.4    0.666667  1.111111      0.04         1.2  
4                  0.7      0.6    1.000000  1.428571      0.18         inf  
5                  0.6      0.6    0.857143  1.428571      0.18         2.8  
6                  0.6      0.4    0.666667  1.111111      0.04         1.2  
7                  0.6      0.4    0.800000  1.333333      0.10         2.0  
8                  0.7      0.4    1.000000  1.428571      0.12         inf  
9                  0.4      0.4    0.571429  1.428571      0.12         1.4  
10                 0.5      0.4    0.666667  1.333333      0.10         1.5  
11                 0.6      0.4    0.666667  1.111111      0.04         1.2

In [23]:
rules3_1 = association_rules(frequent_itemsets3, metric="lift", min_threshold=0.2)
rules3_1

antecedents               consequents  antecedent support  \
0                (Gladiator)             (Sixth Sense)                 0.7   
1              (Sixth Sense)               (Gladiator)                 0.6   
2                  (Patriot)             (Sixth Sense)                 0.6   
3              (Sixth Sense)                 (Patriot)                 0.6   
4                  (Patriot)               (Gladiator)                 0.6   
5                (Gladiator)                 (Patriot)                 0.7   
6       (Patriot, Gladiator)             (Sixth Sense)                 0.6   
7   (Sixth Sense, Gladiator)                 (Patriot)                 0.5   
8     (Patriot, Sixth Sense)               (Gladiator)                 0.4   
9                (Gladiator)    (Patriot, Sixth Sense)                 0.7   
10                 (Patriot)  (Sixth Sense, Gladiator)                 0.6   
11             (Sixth Sense)      (Patriot, Gladiator)                 0.6   

    consequent support  support  confidence      lift  leverage  conviction  
0                  0.6      0.5    0.714286  1.190476      0.08         1.4  
1                  0.7      0.5    0.833333  1.190476      0.08         1.8  
2                  0.6      0.4    0.666667  1.111111      0.04         1.2  
3                  0.6      0.4    0.666667  1.111111      0.04         1.2  
4                  0.7      0.6    1.000000  1.428571      0.18         inf  
5                  0.6      0.6    0.857143  1.428571      0.18         2.8  
6                  0.6      0.4    0.666667  1.111111      0.04         1.2  
7                  0.6      0.4    0.800000  1.333333      0.10         2.0  
8                  0.7      0.4    1.000000  1.428571      0.12         inf  
9                  0.4      0.4    0.571429  1.428571      0.12         1.4  
10                 0.5      0.4    0.666667  1.333333      0.10         1.5  
11                 0.6      0.4    0.666667  1.111111      0.04         1.2

In [24]:
# Creating rules with min_support = 0.5
frequent_itemsets4 = apriori(df, min_support=0.5, use_colnames=True)
frequent_itemsets4

/opt/anaconda3/lib/python3.9/site-packages/mlxtend/frequent_patterns/fpcommon.py:111: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


support                  itemsets
0      0.6             (Sixth Sense)
1      0.7               (Gladiator)
2      0.6                 (Patriot)
3      0.5  (Gladiator, Sixth Sense)
4      0.6      (Patriot, Gladiator)

In [25]:
rules4_1 = association_rules(frequent_itemsets4, metric="lift", min_threshold=0.2)
rules4_1

antecedents    consequents  antecedent support  consequent support  \
0    (Gladiator)  (Sixth Sense)                 0.7                 0.6   
1  (Sixth Sense)    (Gladiator)                 0.6                 0.7   
2      (Patriot)    (Gladiator)                 0.6                 0.7   
3    (Gladiator)      (Patriot)                 0.7                 0.6   

   support  confidence      lift  leverage  conviction  
0      0.5    0.714286  1.190476      0.08         1.4  
1      0.5    0.833333  1.190476      0.08         1.8  
2      0.6    1.000000  1.428571      0.18         inf  
3      0.6    0.857143  1.428571      0.18         2.8

In [26]:
rules4_1 = association_rules(frequent_itemsets4, metric="lift", min_threshold=0.1)
rules4_1

antecedents    consequents  antecedent support  consequent support  \
0    (Gladiator)  (Sixth Sense)                 0.7                 0.6   
1  (Sixth Sense)    (Gladiator)                 0.6                 0.7   
2      (Patriot)    (Gladiator)                 0.6                 0.7   
3    (Gladiator)      (Patriot)                 0.7                 0.6   

   support  confidence      lift  leverage  conviction  
0      0.5    0.714286  1.190476      0.08         1.4  
1      0.5    0.833333  1.190476      0.08         1.8  
2      0.6    1.000000  1.428571      0.18         inf  
3      0.6    0.857143  1.428571      0.18         2.8

## Summary
#### When used min_support as 0.1 and min_threshold as 0.7, 0.1, 0.2 we got around 250 rules
#### when used min_support as 0.2 and min_threshold as 0.7, 0.1, 0.2 we got around 15 rules
#### when used min_support as 0.3 and min_threshold as 0.1, 0.2 we got around 11 rules
#### when used min_support as 0.4 and min_threshold as 0.1, 0.2 we got around 11 rules
#### when used min_support as 0.5 and min_threshold as 0.1, 0.2 we got around 4 rules